### PR TITLE
PYTHON-2504 Apply one time pyupgrade --py37-plus v3.3.2 and ruff

### PR DESCRIPTION
**Pull request by ShaneHarvey**
https://jira.mongodb.org/browse/PYTHON-2504

Successor to https://github.com/mongodb/mongo-python-driver/pull/882.

Ran this repeatedly until it stablized after 3 or 4 runs:
```
$ pyupgrade --py37-plus bson/*.py pymongo/*.py gridfs/*.py test/*.py tools/*.py test/*/*.py
```

I also ran ruff 0.0.265 like this (see https://beta.ruff.rs/docs/rules/):
```
$ ruff --fix-only --select ALL --fixable ALL --target-version py37 --line-length=100 --unfixable COM812,D400,D415,ERA001,RUF100,SIM108,D211,D212,SIM105,SIM,PT,ANN204,EM bson/*.py pymongo/*.py gridfs/*.py test/*.py test/*/*.py
```

The --unfixable codes suppress some unneeded or particularly noisy changes.

As noted in the previous PR:
> I tested out adding this to pre-commit but pyupgrade's formatting conflicts with black so it's more of a nuisance.

I'll add this to `.git-blame-ignore-revs` after the PR is merged.

### CODE REVIEW

**Comment by ShaneHarvey** [comment]
I'm waiting until we get mypy working to mark this as ready.

**Comment by blink1073** [comment]
The `mypy --strict` is failing, not sure why it is being so strict.  Perhaps we need to tell it to ignore imports for `pymongo` and `bson` and focus just on the code in the test?

**Comment by ShaneHarvey** [comment]
I'm going to back out the mypy changes for this PR and open a new ticket.

**Comment by ShaneHarvey** [comment]
I reopened https://jira.mongodb.org/browse/PYTHON-3679 to track mypy --strict support.

**Review by blink1073 — APPROVED** [review]
LGTM!

**Comment by ShaneHarvey** [comment]
I'm going to wait until https://github.com/mongodb/mongo-python-driver/pull/1138 is merged to avoid adding merge conflicts to that PR.

**Comment by ShaneHarvey** [comment]
Rebased and reran the upgrade commands. Waiting for tests to pass.

File: bson/__init__.py [hunk 1]
```diff
@@ -269,10 +267,12 @@ def _get_string(
     length = _UNPACK_INT_FROM(data, position)[0]
     position += 4
     if length < 1 or obj_end - position < length:
-        raise InvalidBSON("invalid string length")
+        msg = "invalid string length"
+        raise InvalidBSON(msg)
```
Comment (ShaneHarvey) [inline bson/__init__.py:271]: Ew I don't like what ruff is doing here at all. Looking for a code to disable.

Comment (blink1073) [reply bson/__init__.py:271]: That rule comes from https://pypi.org/project/flake8-errmsg/

Comment (blink1073) [reply bson/__init__.py:271]: https://beta.ruff.rs/docs/rules/#flake8-errmsg-em

Comment (ShaneHarvey) [reply bson/__init__.py:271]: Caused by https://beta.ruff.rs/docs/rules/#flake8-errmsg-em

Disabling now.

**Review by blink1073 — APPROVED** [review]
I did some spot checks, LGTM!

**Pull request merged**